src/rfq/protocols/hashflow/state.rs

```diff
@@ -37,15 +37,47 @@ impl HashflowState {
     ) -> Self {
         Self { base_token, quote_token, levels, client }
     }
+
+    fn valid_direction_guard(
+        &self,
+        token_address_in: &Bytes,
+        token_address_out: &Bytes,
+    ) -> Result<(), SimulationError> {
+        // The current levels are only valid for the base/quote pair.
+        if !(token_address_in == &self.base_token.address &&
+            token_address_out == &self.quote_token.address)
+        {
+            Err(SimulationError::InvalidInput(
+                format!("Invalid token addresses: {token_address_in}, {token_address_out}"),
+                None,
+            ))
+        } else {
+            Ok(())
+        }
+    }
+
+    fn valid_levels_guard(&self) -> Result<(), SimulationError> {
+        if self.levels.levels.is_empty() {
+            return Err(SimulationError::RecoverableError("No liquidity".into()));
+        }
+        Ok(())
+    }
 }
 
 impl ProtocolSim for HashflowState {
     fn fee(&self) -> f64 {
         todo!()
     }
 
-    fn spot_price(&self, _base: &Token, _quote: &Token) -> Result<f64, SimulationError> {
-        todo!()
+    fn spot_price(&self, base: &Token, quote: &Token) -> Result<f64, SimulationError> {
+        self.valid_direction_guard(&base.address, &quote.address)?;
+
+        // Hashflow's levels are sorted by price, so the first level represents the best price.
+        self.levels
+            .levels
+            .first()
+            .ok_or(SimulationError::RecoverableError("No liquidity".into()))
+            .map(|level| level.price)
```

Comment:
I think this is fine - Bebop also sorted by price leaving the best one first! So no need to do the work twice by finding the min :)